executable-monitor/executable-monitor.py

```diff
@@ -72,76 +81,114 @@
-
-        # Sleep for a short duration between loops to not steal all system resources
-        time.sleep(.1)
-
-    if not exe_exitted:
-        exe.kill()
-
-    # Capture remaining output and check for the successful line
-    for exe_stdout_line in exe.stdout.readlines():
-        logging.info(exe_stdout_line)
-        if args.success_line is not None and args.success_line in exe_stdout_line:
-            success_line_found = True
-    
-    logging.info("END OF DEVICE OUTPUT\n")
-
-    logging.info("EXECUTABLE RUN SUMMARY:\n")
+        timeout_time_seconds = cur_time_seconds + args.timeout_seconds
+
+        logging.info("START OF DEVICE OUTPUT\n")
+
+        # While a timeout hasn't happened, the executable is running, and an exit condition has not been met
+        while ( not exit_condition_met ):
+            # Uncomment this sleep for a short duration between loops to not steal all system resources
+            # time.sleep(.05)
+
+            # Check if executable exitted
+            exe_exit_status = exe.poll()
+            if exe_exit_status is not None:
+                logging.info(f"EXECUTABLE CLOSED WITH STATUS: {exe_exit_status}")
+                exe_exitted = True
+                exit_condition_met = True
+
+            # Read executable's stdout and write to stdout and logfile
+            # A potential improvement here would be to do readlines() on the file, then truncate()
+            # This might be cleaner than this approach of reading a single line each loop.
+            exe_stdout_line = ReadOutputFile.readline()
+            if(exe_stdout_line is not None) and (len(exe_stdout_line.strip()) > 1):
+                # Check if the executable printed out its success line
+                if ( args.success_line is not None ) and ( args.success_line in exe_stdout_line ) :
+                    logging.info(f"SUCCESS_LINE_FOUND: {exe_stdout_line}")
+                    success_line_found = True
+                    success_line = exe_stdout_line
+                    if( not wait_for_exit ):
+                        exit_condition_met = True
+                else:
+                    logging.info(exe_stdout_line)
+
+            # Check for timeout
+            cur_time_seconds = time.time()
+            if cur_time_seconds >= timeout_time_seconds:
+                logging.info(f"TIMEOUT OF {args.timeout_seconds} SECONDS HIT\n")
+                timeout_occurred = True
+                exit_condition_met = True
+
+        if not exe_exitted:
+            logging.info(f"EXECUTABLE DID NOT EXIT, MANUALLY KILLING NOW\n")
+            exe.kill()
+
+        if not exit_condition_met:
+            logging.info(f"PARSING REST OF LOG\n")
+            # Capture remaining output and check for the successful line
+            for exe_stdout_line in ReadOutputFile.readlines():
+                logging.info(exe_stdout_line)
+                if args.success_line is not None and args.success_line in exe_stdout_line:
```

Comment:
Should success_line be a regex rather than a particular line?

Reply:
Not a bad idea, but think for now want to stick with this instead of the regex. Can add that in down the line if we think it would be useful.